underfs/tos/src/main/java/alluxio/underfs/tos/TOSUnderFileSystem.java

```diff
@@ -0,0 +1,335 @@
+    if (output != null) {
+      return new TOSObjectListingChunk(input, output);
+    }
+    return null;
+  }
+
+  // Get next chunk of listing result
+  private ListObjectsType2Output getObjectListingChunk(ListObjectsType2Input input) {
+    ListObjectsType2Output result;
+    try {
+      result = mClient.listObjectsType2(input);
+    } catch (TosException e) {
+      LOG.error("Failed to list path {}", input.getPrefix(), e);
+      result = null;
+    }
+    return result;
+  }
+
+  /**
+   * Wrapper over TOS {@link ObjectListingChunk}.
+   */
+  private final class TOSObjectListingChunk implements ObjectListingChunk {
+    final ListObjectsType2Input mInput;
+    final ListObjectsType2Output mOutput;
+
+    TOSObjectListingChunk(ListObjectsType2Input Input, ListObjectsType2Output Output)
+        throws IOException {
+      mInput = Input;
+      mOutput = Output;
+      if (mOutput == null) {
+        throw new IOException("TOS listing result is null");
+      }
+    }
+
+    @Override
+    public ObjectStatus[] getObjectStatuses() {
+      List<ListedObjectV2> objects = mOutput.getContents();
+      if (objects == null) {
+        return new ObjectStatus[0];
+      }
+      ObjectStatus[] ret = new ObjectStatus[objects.size()];
+      int i = 0;
+      for (ListedObjectV2 obj : objects) {
+        Date lastModifiedDate = obj.getLastModified();
+        Long lastModifiedTime = lastModifiedDate == null ? null : lastModifiedDate.getTime();
+        ret[i++] = new ObjectStatus(obj.getKey(), obj.getEtag(), obj.getSize(),
+            lastModifiedTime);
+      }
+      return ret;
+    }
+
+    @Override
+    public String[] getCommonPrefixes() {
+      List<ListedCommonPrefix> res = mOutput.getCommonPrefixes();
+      if (res == null) {
+        return new String[0];
+      }
+      return res.stream().map(ListedCommonPrefix::getPrefix).toArray(String[]::new);
+    }
+
+    @Override
+    public ObjectListingChunk getNextChunk() throws IOException {
+      if (mOutput.isTruncated()) {
+        mInput.setContinuationToken(mOutput.getNextContinuationToken());
+        ListObjectsType2Output nextResult = mClient.listObjectsType2(mInput);
+        if (nextResult != null) {
+          return new TOSObjectListingChunk(mInput, nextResult);
+        }
+      }
+      return null;
+    }
+  }
+
+  @Override
+  protected ObjectStatus getObjectStatus(String key) {
+    try {
+      HeadObjectV2Input input = new HeadObjectV2Input().setBucket(mBucketName).setKey(key);
+      HeadObjectV2Output output = mClient.headObject(input);
+      if (output == null) {
+        return null;
+      }
+      Date lastModifiedDate = output.getLastModifiedInDate();
+      Long lastModifiedTime = lastModifiedDate == null ? null : lastModifiedDate.getTime();
+      return new ObjectStatus(key, output.getEtag(), output.getContentLength(),
+          lastModifiedTime);
+    } catch (TosException e) {
+      return null;
```

Comment:
please check the exception, if it's 404 not found, then return null, otherwise throw a exception